### PR TITLE
[perf] Only sort relocations once before object file emission

### PR DESCRIPTION
**Pull request by UplinkCoder**
Before this patch the backend would run an insertion sort,
every time a new relocation was added which has O(N^2) behavior
this is currently the biggest bottleneck when generating code with
a huge number of symbols.

cc @WalterBright

### CODE REVIEW

**Comment by rikkimax** [comment]
Do you have some comparative benchmarks for this?

**Comment by UplinkCoder** [comment]
@rikkimax yes for uses core.reflect it cuts the code-emission time from 650 milliseconds to 70 milliseconds

**Review by RazvanN7 — APPROVED** [review]

**Comment by UplinkCoder** [comment]
Instead of calling qsort we _could_ also do 'Patience sorting' which would be actually optimal for this usecase. But that would mean keeping around another auxilary array whose state has to be managed.
And implementing the sort has the possibility of introducing bugs.

File: src/dmd/backend/elfobj.d [hunk 1]
```diff
@@ -1195,18 +1198,47 @@ void ElfObj_term(const(char)* objfilename)
...
-                        assert(*cast(Elf64_Xword *)(seg.SDbuf.buf + p.r_offset) == 0);
-                }
+                extern (C) @trusted nothrow
+                    static int elf64_rel_fp(scope const(void*) e1,
+                                            scope const(void*) e2)
+                    {
+                        Elf64_Rela *r1 = cast(Elf64_Rela *)e1;
+                        Elf64_Rela *r2 = cast(Elf64_Rela *)e2;
+
+                        return cast(int)(r1.r_offset - r2.r_offset);
```
Comment (WalterBright) [inline src/dmd/backend/elfobj.d:1213]: I'm concerned about wraparound here when only the lower 32 bits are compared.

Comment (WalterBright) [reply src/dmd/backend/elfobj.d:1213]: Better if compare the full size of r_offset.

Comment (dkorpel) [reply src/dmd/backend/elfobj.d:1213]: Druntime likes to use this pattern I believe:
```suggestion
                        return (r1.r_offset > r2.r_offset) - (r1.r_offset < r2.r_offset); 
```

Comment (UplinkCoder) [reply src/dmd/backend/elfobj.d:1213]: I like this.

File: src/dmd/backend/elfobj.d [hunk 2]
```diff
@@ -1195,18 +1198,47 @@ void ElfObj_term(const(char)* objfilename)
...
-                        assert(*cast(Elf64_Xword *)(seg.SDbuf.buf + p.r_offset) == 0);
-                }
+                extern (C) @trusted nothrow
+                    static int elf64_rel_fp(scope const(void*) e1,
+                                            scope const(void*) e2)
+                    {
+                        Elf64_Rela *r1 = cast(Elf64_Rela *)e1;
+                        Elf64_Rela *r2 = cast(Elf64_Rela *)e2;
+
+                        return (r1.r_offset > r2.r_offset) - (r1.r_offset < r2.r_offset);
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d:1213]: @UplinkCoder there is another cast here

Comment (thewilsonator) [reply src/dmd/backend/elfobj.d:1213]: whoops wrong place.

File: src/dmd/backend/elfobj.d [hunk 3]
```diff
@@ -1195,18 +1198,47 @@ void ElfObj_term(const(char)* objfilename)
...
+            {
                 assert(seg.SDrelcnt == seg.SDrel.length() / Elf32_Rel.sizeof);
+                extern (C) @trusted nothrow
+                    static int elf32_rel_fp(scope const(void*) e1,
+                                            scope const(void*) e2)
+                    {
+                        Elf32_Rel *r1 = cast(Elf32_Rel *)e1;
+                        Elf32_Rel *r2 = cast(Elf32_Rel *)e2;
+
+                        return cast(int)(r1.r_offset - r2.r_offset);
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d:1232]: I meant here.

**Comment by thewilsonator** [comment]
`/dmd/src/dmd/backend/elfobj.d' contains trailing whitespace at line 1213`

File: src/dmd/backend/elfobj.d (hunk 3 above)
Comment (UplinkCoder) [reply src/dmd/backend/elfobj.d:1232]: 32bit offset.
that works

Comment (thewilsonator) [reply src/dmd/backend/elfobj.d:1232]: OK

Comment (rainers) [reply src/dmd/backend/elfobj.d]: > 32bit offset. that works

not if the difference of the `uint`s is larger than `1<<31`

Comment (UplinkCoder) [reply src/dmd/backend/elfobj.d]: I changed the code already before it was merged.

**Pull request merged**